### PR TITLE
Update digest-sha on any source change

### PR DESCRIPTION
**Pull request by patrickbkr**
The nqp-config.nqp file contains a checksum of all the source code. That
checksum needs to be up-to-date for a logic in Rakudo that forces a rebuild
depending on changes in NQP to work. So make sure the checksum is
regenerated whenever the sources change.

### CODE REVIEW

**Comment by lizmat** [comment]
Didn't look at the patch.  Possibly related: while working on the rakuast branch, I need to always do a full reconfigure whenever I change something in `src/Raku/ast`.  Which takes another 20 seconds or so additional roundtrip  :-(  .  Could you have a look there if there's an easy fix for that?

**Comment by patrickbkr** [comment]
@lizmat What's the error you get when not doing the reconfigure?

**Comment by lizmat** [comment]
$ make
    +++ Checking for moar NQP version
    '/Users/liz/Github/rakuast/install/bin/nqp-m' tools/build/raku-ast-compiler.nqp src/Raku/ast/impl.rakumod src/Raku/ast/resolver.rakumod src/Raku/ast/base.rakumod src/Raku/ast/sink.rakumod src/Raku/ast/scoping.rakumod src/Raku/ast/attaching.rakumod src/Raku/ast/begintime.rakumod src/Raku/ast/traits.rakumod src/Raku/ast/statements.rakumod src/Raku/ast/expressions.rakumod src/Raku/ast/circumfix.rakumod src/Raku/ast/pair.rakumod src/Raku/ast/call.rakumod src/Raku/ast/term.rakumod src/Raku/ast/name.rakumod src/Raku/ast/type.rakumod src/Raku/ast/meta.rakumod src/Raku/ast/variable-declaration.rakumod src/Raku/ast/variable-access.rakumod src/Raku/ast/compunit.rakumod src/Raku/ast/code.rakumod src/Raku/ast/statementprefixes.rakumod src/Raku/ast/signature.rakumod src/Raku/ast/package.rakumod src/Raku/ast/literals.rakumod src/Raku/ast/regex.rakumod > gen/ast.nqp
    +++ Generating	gen/moar/BOOTSTRAP/v6c.nqp
    +++ Compiling	blib/Perl6/BOOTSTRAP/v6c.moarvm
    +++ Compiling	blib/CORE.c.setting.moarvm
    The following step can take a long time, please be patient.
    Stage start      :   0.000
    Stage parse      :  49.178
    Stage syntaxcheck:   0.000
    Stage ast        :   0.000
    Stage optimize   :   5.003
    Stage mast       :  10.757
    Stage mbc        :   2.516
    +++ Generating	rakudo-m
    Missing or wrong version of dependency 'gen/moar/BOOTSTRAP/v6c.nqp' (from 'gen/moar/CORE.d.setting')
       at <unknown>:1  (/Users/liz/Github/rakuast/blib/CORE.d.setting.moarvm:<dependencies+deserialize>)
     from gen/moar/ModuleLoader.nqp:274  (/Users/liz/Github/rakuast/blib/Perl6/ModuleLoader.moarvm:)
     from gen/moar/ModuleLoader.nqp:269  (/Users/liz/Github/rakuast/blib/Perl6/ModuleLoader.moarvm:)
     from gen/moar/ModuleLoader.nqp:258  (/Users/liz/Github/rakuast/blib/Perl6/ModuleLoader.moarvm:load_setting)
     from gen/moar/World.nqp:958  (/Users/liz/Github/rakuast/blib/Perl6/World.moarvm:load_setting)
     from gen/moar/World.nqp:710  (/Users/liz/Github/rakuast/blib/Perl6/World.moarvm:comp_unit_stage1)
     from gen/moar/Grammar.nqp:840  (/Users/liz/Github/rakuast/blib/Perl6/Grammar.moarvm:comp_unit)
     from gen/moar/Grammar.nqp:555  (/Users/liz/Github/rakuast/blib/Perl6/Grammar.moarvm:TOP)
     from gen/moar/stage2/QRegex.nqp:2258  (/Users/liz/Github/rakuast/install/share/nqp/lib/QRegex.moarvm:parse)
     from gen/moar/stage2/NQPHLL.nqp:2297  (/Users/liz/Github/rakuast/install/share/nqp/lib/NQPHLL.moarvm:parse)
     from gen/moar/stage2/NQPHLL.nqp:2213  (/Users/liz/Github/rakuast/install/share/nqp/lib/NQPHLL.moarvm:execute_stage)
     from gen/moar/stage2/NQPHLL.nqp:2248  (/Users/liz/Github/rakuast/install/share/nqp/lib/NQPHLL.moarvm:run)
     from gen/moar/stage2/NQPHLL.nqp:2244  (/Users/liz/Github/rakuast/install/share/nqp/lib/NQPHLL.moarvm:)
     from gen/moar/stage2/NQPHLL.nqp:2240  (/Users/liz/Github/rakuast/install/share/nqp/lib/NQPHLL.moarvm:compile)
     from gen/moar/stage2/NQPHLL.nqp:1915  (/Users/liz/Github/rakuast/install/share/nqp/lib/NQPHLL.moarvm:eval)
     from gen/moar/stage2/NQPHLL.nqp:2150  (/Users/liz/Github/rakuast/install/share/nqp/lib/NQPHLL.moarvm:evalfiles)
     from gen/moar/stage2/NQPHLL.nqp:2110  (/Users/liz/Github/rakuast/install/share/nqp/lib/NQPHLL.moarvm:command_eval)
     from gen/moar/Compiler.nqp:64  (/Users/liz/Github/rakuast/blib/Perl6/Compiler.moarvm:command_eval)
     from gen/moar/stage2/NQPHLL.nqp:2035  (/Users/liz/Github/rakuast/install/share/nqp/lib/NQPHLL.moarvm:command_line)
     from gen/moar/rakudo.nqp:134  (rakudo.moarvm:MAIN)
     from gen/moar/rakudo.nqp:99  (rakudo.moarvm:<mainline>)
     from <unknown>:1  (rakudo.moarvm:<main>)
     from <unknown>:1  (rakudo.moarvm:<entry>)
    make: *** [RESTRICTED.setting.moarvm] Error 1

File: tools/templates/Makefile-backend-common.in [hunk 1]
```diff
@@ -118,7 +118,7 @@
 @stage_precomp(@bsm(P6QREGEX_@ucstage@)@:					@use_prereqs(@bpm(P6QREGEX_COMBINED_@ucstage@)@)@ @bsm(@ucstage@_DEPS)@ @bsm(HLL_@ucstage@)@ @bsm(QREGEX_@ucstage@)@ @bsm(QAST_@ucstage@)@ @bsm(QASTNODE_@ucstage@)@)@
 
 
-@nfp(@stage_dir@/nqp-config.nqp)@:
+@nfp(@stage_dir@/nqp-config.nqp)@: @bpm(SOURCES)@
```
Comment (niner) [inline tools/templates/Makefile-backend-common.in:121]: Previously this recipie was part of the following target and depended on `@bsm(NQP_@ucprev_stage@)@ @nfp(@stage_dir@/@bsm(QAST)@)@ @nfp(@stage_dir@/@bsm(P6QREGEX)@)@ @bpm(SOURCES)@`

Does `@bpm(SOURCES)@` already contain all `.nqp` files from `src/`?

Btw. now that I see this, I think we have always missed `MAST/Ops.nqp` and `MAST/Nodes.nqp` in `gen-version.pl`

**Review by niner — APPROVED** [review]

File: tools/templates/Makefile-backend-common.in (hunk 1 above)
Comment (patrickbkr) [reply tools/templates/Makefile-backend-common.in:121]: On moar it does with the exception of `P6QREGEX_SOURCES`. I'm not sure if that's missing on purpose.
On JVM it includes everything.
I didn't check on JS.

**Review by niner — APPROVED** [review]
P6QREGEX_SOURCES was not needed in COMMON_SOURCES because the NQP_MOAR targets already depend on P6QREGEX_MOAR, i.e. the compilation result. Adding it won't hurt though and could prevent future mistakes there, so I think it's OK.

**Pull request merged**